File: Merge_Sort/Merge_Sort.py
# Merge sort in python

def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        left_half = arr[:mid]
        right_half = arr[mid:]

        # Recursively sort both halves
        merge_sort(left_half)
        merge_sort(right_half)

        # Merge the sorted halves
        i = j = k = 0
        while i < len(left_half) and j < len(right_half):
            if left_half[i] < right_half[j]:
                arr[k] = left_half[i]
                i += 1
            else:
                arr[k] = right_half[j]
                j += 1
            k += 1

        # Check for any remaining elements
        while i < len(left_half):
            arr[k] = left_half[i]
            i += 1
            k += 1
        while j < len(right_half):
            arr[k] = right_half[j]
            j += 1
            k += 1
    return arr


#Example
data=[-2,45,0,11,-9]
print("Sorted Array in Ascending Order: ")
print(merge_sort(data))

# This code is contributed by Erfan Shafiee

File: Merge_Sort/Merge_Sort.cs
// Merge sort in C#
using System;

public class Merge_Sort
{
    // Merge two subarrays L and M into arr
    static void Merge(int[] arr, int p, int q, int r)
    {
        // Create L ← A[p..q] and M ← A[q+1..r]
        int n1 = q - p + 1;
        int n2 = r - q;

        int[] L = new int[n1];
        int[] M = new int[n2];

        for (int i = 0; i < n1; i++)
            L[i] = arr[p + i];
        for (int j = 0; j < n2; j++)
            M[j] = arr[q + 1 + j];

        // Maintain current index of sub-arrays and main array
        int i, j, k;
        i = 0;
        j = 0;
        k = p;

        // Until we reach either end of either L or M, pick larger among
        // elements L and M and place them in the correct position at A[p..r]
        while (i < n1 && j < n2)
        {
            if (L[i] <= M[j])
            {
                arr[k] = L[i];
                i++;
            }
            else
            {
                arr[k] = M[j];
                j++;
            }
            k++;
        }

        // When we run out of elements in either L or M,
        // pick up the remaining elements and put in A[p..r]
        while (i < n1)
        {
            arr[k] = L[i];
            i++;
            k++;
        }

        while (j < n2)
        {
            arr[k] = M[j];
            j++;
            k++;
        }
    }

    // Divide the array into two subarrays, sort them and merge them
    static void MergeSort(int[] arr, int l, int r)
    {
        if (l < r)
        {
            // m is the point where the array is divided into two subarrays
            int m = l + (r - l) / 2;

            MergeSort(arr, l, m);
            MergeSort(arr, m + 1, r);

            // Merge the sorted subarrays
            Merge(arr, l, m, r);
        }
    }

    // Print array function
    static void PrintArray(int[] arr, int size)
    {
        for (int i = 0; i < size; i++)
            Console.Write(arr[i] + " ");
        Console.WriteLine();
    }

    // Main function
    public static void Main(string[] args)
    {
        int[] arr = { 6, 5, 12, 10, 9, 1 };
        int size = arr.Length;

        MergeSort(arr, 0, size - 1);

        Console.WriteLine("Sorted array: ");
        PrintArray(arr, size);
    }
}

// This code is contributed by Ali Barzegari Dahaj